sklearn/neighbors/_nearest_centroid.py

```diff
@@ -255,7 +296,95 @@ def predict(self, X):
+        )
+
+        for cur_class in range(self.classes_.size):
+            Xdist = X - self.centroids_[cur_class, :]
+            Xdist_norm = np.square(Xdist / self.s_)
+            # Hastie et al. (2009), p. 652, Eq. (18.2)
+            discriminant_score[:, cur_class] = np.squeeze(
+                -np.sum(Xdist_norm, axis=1) + 2.0 * np.log(self.priors_[cur_class])
+            )
+
+        return discriminant_score
+
+    def decision_function(self, X):
+        """Apply decision function to an array of samples.
+
+        The decision function is equal (up to a constant factor) to the
+        log-posterior of the model, i.e. `log p(y = k | x)`.
+        However, the decision function uses the argmax definition instead
+        of argmin definition as in the predict() function, which are just
+        -1 factor of each other.
+        The estimation has been implemented according to
+        Hastie et al. (2009), p. 652 equation (18.2)
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            Array of samples (test vectors).
+
+        Returns
+        -------
+        C : ndarray of shape (n_samples,) or (n_samples, n_classes)
+            Decision function values related to each class, per sample.
+            In the two-class case, the shape is (n_samples,), giving the
+            log likelihood ratio of the positive class.
+        """
+
+        dec_func = self._decision_function(X)
+        # handle special case of two classes
+        if len(self.classes_) == 2:
+            return dec_func[:, 1] - dec_func[:, 0]
+        return dec_func
+
+    def predict_proba(self, X):
+        """Estimate class probabilities.
+
+        The returned estimates for all classes are ordered by the
+        label of classes. The estimation has been implemented according to
+        Hastie et al. (2009), p. 652 equation (18.8).
+
+        Parameters
+        ----------
+        X : array-like, shape = [n_samples, n_features]
+            Test samples.
+
+        Returns
+        -------
+        T : array-like, shape = [n_samples, n_classes]
+            Returns the probability of the sample for each class in the model,
+            where classes are ordered as they are in ``self.classes_``.
+        """
+        coef = self._decision_function(X)
+        likelihood = np.exp(coef - coef.max(axis=1)[:, np.newaxis])
+        return likelihood / likelihood.sum(axis=1)[:, np.newaxis]
+
+    def predict_log_proba(self, X):
+        """Estimate log probability.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            Test samples.
+
+        Returns
+        -------
+        C : ndarray of shape (n_samples, n_classes)
+            Estimated log probabilities.
+        """
+        prediction = self.predict_proba(X)
+        prediction[prediction == 0.0] += np.finfo(prediction.dtype).tiny
+        return np.log(prediction)
```

Comment:
I have to check more but we need to put in common this code with the one of the `LinearDiscriminantAnalysis` or `QuadraticDiscriminantAnalysis`. I have to check which definition is equivalent but from the paper, it seems that this is exactly the same defitnion than the Gaussian linear discriminant analysis.

Reply:
So these are identical to those in `QuadraticDiscriminantAnalysis`. Should we treat them as separate issue/PR? And where do you suggest putting these functions? Right now, each classifier has its own implementation of _predict_proba_ and _predict_log_proba_.

Reply:
> Should we treat them as separate issue/PR?

I would do it in this PR. I would suggest to create a pure Python function in the discriminant_analysis module and the method from both estimator will call the function. We probably only need a single function to convert the decision function to probabilities. The log proba is quite straight forward and does not need to share code.